qiskit/circuit/library/standard_gates/x.py

```diff
@@ -1188,60 +1196,132 @@ def _define(self):
         definition = []
 
         if self._dirty_ancillas:
-            i = self.num_ctrl_qubits - 3
-            ancilla_pre_rule = [
-                (U2Gate(0, numpy.pi), [q_target], []),
-                (CXGate(), [q_target, q_ancillas[i]], []),
-                (U1Gate(-numpy.pi / 4), [q_ancillas[i]], []),
-                (CXGate(), [q_controls[-1], q_ancillas[i]], []),
-                (U1Gate(numpy.pi / 4), [q_ancillas[i]], []),
-                (CXGate(), [q_target, q_ancillas[i]], []),
-                (U1Gate(-numpy.pi / 4), [q_ancillas[i]], []),
-                (CXGate(), [q_controls[-1], q_ancillas[i]], []),
-                (U1Gate(numpy.pi / 4), [q_ancillas[i]], []),
-            ]
-            for inst in ancilla_pre_rule:
-                definition.append(inst)
-
-            for j in reversed(range(2, self.num_ctrl_qubits - 1)):
+            if self.num_ctrl_qubits < 3:
+                definition.append((MCXGate(self.num_ctrl_qubits), [*q_controls, q_target], []))
+            elif not self._relative_phase and self.num_ctrl_qubits == 3:
+                definition.append((C3XGate(), [*q_controls, q_target], []))
+            else:
+                num_ancillas = self.num_ctrl_qubits - 2
+                targets = [q_target] + q_ancillas[:num_ancillas][::-1]
+
+                for j in range(2):
+                    for i in range(self.num_ctrl_qubits):  # action part
+                        if i < self.num_ctrl_qubits - 2:
+                            if targets[i] != q_target or self._relative_phase:
+                                # gate cancelling
+
+                                # cancel rightmost gates of action part
+                                # with leftmost gates of reset part
+                                if self._relative_phase and targets[i] == q_target and j == 1:
+                                    definition.append(
+                                        (
+                                            CXGate(),
+                                            [q_ancillas[num_ancillas - i - 1], targets[i]],
+                                            [],
+                                        )
+                                    )
+                                    definition.append((U1Gate(pi / 4), [targets[i]], []))  # T gate
```

Comment:
In the paper by Iten you're referencing, the decomposition of Lemma 8 is using an Ry(pi/4) gate (what they call A), but you're using single T gates sometimes -- could you briefly explain why this is equivalent here?

Reply:
Both result in a Toffoli gate up to some phase. The gate A could be a different gate as long as the circuit is its own inverse. We were trying to keep similarities with Qiskit's code and used the same single-qubit gates from the RCCX circuit, but we could change them to Ry gates.

Reply:
Using a T gate (or any RZ-like gate) is better for now as that can be implemented without a physical pulse on the device. Though that probably doesn't matter so much as these MCX decompositions are likely more interesting for fault-tolerance anyways 😄 Could you use a TGate or PhaseGate instead of `U1` though? The latter is legacy IBM specific.